eZ/Publish/API/Repository/Iterator/BatchIterator.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\API\Repository\Iterator;
+
+use Iterator;
+
+final class BatchIterator implements Iterator
+{
+    public const DEFAULT_BATCH_SIZE = 25;
+
+    /** @var \eZ\Publish\API\Repository\Iterator\BatchIteratorAdapter */
+    private $adapter;
+
+    /** @var Iterator|null */
+    private $innerIterator;
+
+    /** @var mixed|null */
+    private $current = null;
+
+    /** @var int */
+    private $position = 0;
+
+    /** @var int */
+    private $batchSize;
+
+    public function __construct(
+        BatchIteratorAdapter $adapter,
+        int $batchSize = self::DEFAULT_BATCH_SIZE
+    ) {
+        $this->adapter = $adapter;
+        $this->batchSize = $batchSize;
+    }
+
+    public function current()
+    {
+        return $this->current;
+    }
+
+    public function next(): void
+    {
+        if ($this->current === null) {
+            $this->innerIterator = $this->adapter->fetch($this->position, $this->batchSize);
```

Comment:
Moving iterator initialization (or re-initialization) into a separate method would allow you to ensure that `IteratorAggregate` (the second half of `Traversable` union) is converted into `Iterator` here.

This would remove the need to reduce the scope of `SearchResult::getIterator()` (currently `Traversable`, this PR limits it to `Iterator`).